src/modules/launcher/PowerLauncher/SettingsReader.cs

```diff
@@ -215,5 +195,52 @@ private static IEnumerable<PowerLauncherPluginSettings> GetDefaultPluginsSetting
                 AdditionalOptions = x.Plugin is ISettingProvider ? (x.Plugin as ISettingProvider).AdditionalOptions : new List<PluginAdditionalOption>(),
             });
         }
+
+        /// <summary>
+        /// Add new plugins and updates additional options for existing ones
+        /// </summary>
+        private static void UpdateSettings(PowerLauncherSettings settings)
+        {
+            var plugins = GetDefaultPluginsSettings().ToDictionary(x => x.Id);
+
+            foreach (var pluginPair in PluginManager.AllPlugins)
+            {
+                var plugin = pluginPair.Plugin;
+                var pluginSettings = settings.Plugins.SingleOrDefault(p => p.Id == pluginPair.Metadata.ID);
+
+                // Add plugin
+                if (pluginSettings == null)
+                {
+                    pluginSettings = plugins[pluginPair.Metadata.ID];
+                    settings.Plugins = settings.Plugins.Append(pluginSettings).OrderBy(p => p.Name);
+                    continue;
+                }
+
+                // Update additional options
+                if (plugin is ISettingProvider && (plugin as ISettingProvider).AdditionalOptions != null)
+                {
+                    foreach (var option in (plugin as ISettingProvider).AdditionalOptions)
+                    {
+                        var pluginSetting = pluginSettings.AdditionalOptions.SingleOrDefault(o => o.Key == option.Key);
+
+                        if (pluginSetting == null)
+                        {
+                            pluginSettings.AdditionalOptions = pluginSettings.AdditionalOptions.Append(option);
+                        }
+                        else
+                        {
+                            pluginSetting.DisplayLabel = option.DisplayLabel;
+                        }
+                    }
+
+                    var optionsToDelete = pluginSettings.AdditionalOptions.Except((plugin as ISettingProvider).AdditionalOptions);
+                    pluginSettings.AdditionalOptions = pluginSettings.AdditionalOptions.Except(optionsToDelete);
+                }
+                else if (!(plugin is ISettingProvider))
+                {
+                    pluginSettings.AdditionalOptions = Enumerable.Empty<PluginAdditionalOption>();
+                }
+            }
+        }
```

Comment:
```suggestion
        private static void UpdateSettings(PowerLauncherSettings settings)
        {
            var defaultPlugins = GetDefaultPluginsSettings().ToDictionary(x => x.Id);
            foreach (PowerLauncherPluginSettings plugin in settings.Plugins)
            {
                if (defaultPlugins.ContainsKey(plugin.Id))
                {
                    var additionalOptions = CombineAdditionalOptions(defaultPlugins[plugin.Id].AdditionalOptions, plugin.AdditionalOptions);
                    defaultPlugins[plugin.Id] = plugin;
                    defaultPlugins[plugin.Id].AdditionalOptions = additionalOptions;
                }
            }

            settings.Plugins = defaultPlugins.Values.ToList();
        }

        private static IEnumerable<PluginAdditionalOption> CombineAdditionalOptions(IEnumerable<PluginAdditionalOption> defaultAdditionalOptions, IEnumerable<PluginAdditionalOption> additionalOptions)
        {
            var defaultOptions = defaultAdditionalOptions.ToDictionary(x => x.Key);
            foreach (var option in additionalOptions)
            {
                if (defaultOptions.ContainsKey(option.Key))
                {
                    defaultOptions[option.Key].DisplayLabel = option.DisplayLabel;
                    defaultOptions[option.Key].Value = option.Value;
                }
            }

            return defaultOptions.Values;
        }
```

Reply:
Even though the previous implementation is correct I think that the proposed one is easier to understand